guide/src/class.md

```diff
@@ -192,6 +192,71 @@ If no method marked with `#[new]` is declared, object instances can only be crea
 
 For arguments, see the [`Method arguments`](#method-arguments) section below.
 
+## Initializer
+
+An initializer implements Python's `__init__` method.
+
+It may be required when it's needed to control an object initalization flow on the Rust code.
+If possible handling this in `__new__` should be preferred, but in some cases, like subclassing native types, overwriting `__init__` might be necessary.
+For example, you define a class that extends `PyDict` and don't want that the original `__init__` method of `PyDict` been called.
+In this case by defining an own `__init__` method it's possible to stop initialization flow.
+
+If you declare an own `__init__` method you may need to call a super class' `__init__` method explicitly like that happens in a regular Python code.
+
+To declare an initializer, you need to define the `__init__` method.
+Like in Python `__init__` must have the `self` receiver as the first argument, followed by the same arguments as the constructor.
+It can either return `()` or `PyResult<()>`.
+
+```rust
+# #![allow(dead_code)]
+# use pyo3::prelude::*;
+# #[cfg(not(Py_LIMITED_API))]
+use pyo3::types::{PyDict, PyTuple};
+
+# #[cfg(not(Py_LIMITED_API))]
+#[pyclass(extends = PyDict)]
+struct MyDict;
+
+# #[cfg(not(Py_LIMITED_API))]
+#[pymethods]
+impl MyDict {
+#   #[allow(unused_variables)]
+    #[new]
+    #[pyo3(signature = (*args, **kwargs))]
+    fn __new__(
+        args: &Bound<'_, PyTuple>,
+        kwargs: Option<&Bound<'_, PyDict>>,
+    ) -> PyResult<Self> {
+        Ok(Self)
+    }
+
+    #[pyo3(signature = (*args, **kwargs))]
+    fn __init__(
+        slf: &Bound<'_, Self>,
+        args: &Bound<'_, PyTuple>,
+        kwargs: Option<&Bound<'_, PyDict>>,
+    ) -> PyResult<()> {
+        // call the super types __init__
+        slf.py_super()?
+            .call_method("__init__", args.to_owned(), kwargs)?;
+        slf.as_super().set_item("my_key", "always insert this key")?;
```

Comment:
I think similarly if a subclass of `MyDict` overrides `__setitem__` then this call wouldn't respect it. Maybe not a relevant concern, could just put a comment above noting that if allowing subclasses it might be better to just use `slf.set_item` to allow normal Python method resolution to play.

(Alternatively we could "fix" #4490, might be hard though.)

Reply:
Yeah, I will add a comment here, but there is the same behavior when writing in pure Python, so I think as an example it's ok.

Reply:
Agreed, I think that's fine